File: Editor/lib/utilities/undo_redo.dart
import 'package:editor/common/mvvm/observable_list.dart';

abstract class IUndoRedo {
  String get name;
  void undo();
  void redo();
}

class UndoRedoAction implements IUndoRedo {
  @override
  late final String name;
  late final Function undoAction;
  late final Function redoAction;

  UndoRedoAction(
      {required this.name, required this.undoAction, required this.redoAction});

  @override
  void undo() => undoAction();

  @override
  void redo() => redoAction();
}

class UndoRedo {
  static final UndoRedo _undoRedo = UndoRedo._internal();

  factory UndoRedo() {
    return _undoRedo;
  }

  UndoRedo._internal();

  final ObservableList<IUndoRedo> redoList = ObservableList();
  final ObservableList<IUndoRedo> undoList = ObservableList();

  void add(IUndoRedo cmd) {
    undoList.add(cmd);
    redoList.clear();
  }

  void undo() {
    if (undoList.isNotEmpty) {
      var cmd = undoList.last;
      undoList.remove(undoList.last);
      cmd.undo();
      redoList.insert(0, cmd);
    }
  }

  void redo() {
    if (redoList.isNotEmpty) {
      var cmd = redoList.first;
      redoList.removeAt(0);
      cmd.redo();
      undoList.add(cmd);
    }
  }

  void resset() {
    redoList.clear();
    undoList.clear();
  }
}

File: Editor/Utilities/UndoRedo.cs
using System.Collections.ObjectModel;

namespace Editor.Utilities
{
    public class UndoRedo
    {
        private bool _registerAction = true;
        private readonly ObservableCollection<IUndoRedo> _redoList = [];
        private readonly ObservableCollection<IUndoRedo> _undoList = [];

        public ReadOnlyObservableCollection<IUndoRedo> RedoList { get; }
        public ReadOnlyObservableCollection<IUndoRedo> UndoList { get; }

        public UndoRedo()
        {
            RedoList = new ReadOnlyObservableCollection<IUndoRedo>(_redoList);
            UndoList = new ReadOnlyObservableCollection<IUndoRedo>(_undoList);
        }

        public void Add(IUndoRedo cmd)
        {
            if (_registerAction)
            {
                _undoList.Add(cmd);
                _redoList.Clear();
            }
        }

        public void Undo()
        {
            if (_undoList.Any())
            {
                var cmd = _undoList.Last();
                _undoList.RemoveAt(_undoList.Count - 1);
                _registerAction = false;
                cmd.Undo();
                _registerAction = true;
                _redoList.Insert(0, cmd);
            }
        }

        public void Redo()
        {
            if (_redoList.Any())
            {
                var cmd = _redoList.First();
                _redoList.RemoveAt(0);
                _registerAction = false;
                cmd.Redo();
                _registerAction = true;
                _undoList.Add(cmd);
            }
        }

        public void Reset()
        {
            _redoList.Clear();
            _undoList.Clear();
        }
    }
}
